BasicSampleKotlin/app/src/main/java/com/example/android/persistence/ui/ProductFragment.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2017, The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.persistence.ui
+
+import android.arch.lifecycle.Observer
+import android.arch.lifecycle.ViewModelProviders
+import android.databinding.DataBindingUtil
+import android.os.Bundle
+import android.support.v4.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+
+import com.example.android.persistence.R
+import com.example.android.persistence.databinding.ProductFragmentBinding
+import com.example.android.persistence.model.Comment
+import com.example.android.persistence.viewmodel.ProductViewModel
+
+class ProductFragment : Fragment() {
+    private lateinit var binding: ProductFragmentBinding
+
+    private lateinit var commentAdapter: CommentAdapter
+
+    private val commentClickCallback = object: CommentClickCallback {
+        override fun onClick(comment: Comment) {
+            // no-op
+        }
+    }
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
+                              savedInstanceState: Bundle?): View? {
+        commentAdapter = CommentAdapter(commentClickCallback)
+
+        // Inflate this data binding layout
+        binding = DataBindingUtil.inflate(inflater, R.layout.product_fragment, container, false)
+
+        // Create and set the adapter for the RecyclerView.
+        binding.commentList.adapter = commentAdapter
+        return binding.root
+    }
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+
+        val factory = ProductViewModel.Factory(
+            requireActivity().application, arguments!!.getInt(KEY_PRODUCT_ID))
+
+        val model = ViewModelProviders.of(this, factory).get<ProductViewModel>(ProductViewModel::class.java)
+
+        binding.productViewModel = model
+
+        subscribeToModel(model)
+    }
+
+    private fun subscribeToModel(model: ProductViewModel) {
+
+        // Observe product data
+        model.observableProduct.observe(this, Observer { productEntity -> model.setProduct(productEntity!!) })
```

Comment:
It seems peculiar that we set the productId on the ViewModel as a constructor argument, but then query the product in the fragment and then set the result of the at query / observer back on the ViewModel - what is the advantage of the current implementation and not having the ViewModel set the product internally?

Reply:
probably because of this: 

![screenshot from 2019-02-19 18-11-53](https://user-images.githubusercontent.com/1117489/53056727-f18cdc00-3471-11e9-9053-1f79900b0634.png)

Reply:
I inherited that from the [original Java version of the sample](https://github.com/googlesamples/android-architecture-components/blob/6da3bf1d4512265196438af9d6f970c75036ecaf/BasicSample/app/src/main/java/com/example/android/persistence/ui/ProductFragment.java#L84-L88).

It seems to update an [ObservableField inside ViewModel](https://github.com/googlesamples/android-architecture-components/blob/6da3bf1d4512265196438af9d6f970c75036ecaf/BasicSample/app/src/main/java/com/example/android/persistence/viewmodel/ProductViewModel.java#L64-L66).

Technically the correct solution would be to expose a LiveData that is ran through `Transformations.switchMap {` which also updates the observable field on change and not just the mediator; then the UI subscribes on *that*.

-----------

I made the PR 9 months ago, but apparently I didn't do anything intrusive. In fact, considering I didn't add my favorite `inline fun <reified T:` function for creating ViewModels, I *really* didn't try to change it up much.

There's a good chance that the BasicSample needs as much look as this one, considering this is just a copy; but then the two should be kept in sync.